### PR TITLE
fix(sglang): fail closed when aux collection captures the last target layer pre-norm

### LINKED ISSUE #20: SGLang aux capture uses pre-norm for the last target layer, diverging from offline / old-logprob (post-norm)
## Summary

The target hidden-state capture for drafter training has a **final-norm semantic divergence** between the SGLang online collection path and both the AutoModel offline reference and verl-SpeCo's own old-logprob collection path. For a `target_layer_id` equal to the target's **last layer** (`L-1`), SGLang captures the **raw pre-norm** residual, while AutoModel and the old-logprob path capture the **post-norm** hidden. This silently gives the drafter's `fc` an inconsistent feature for that slot across collection paths / between offline-train and online-serve.

## Where it comes from

SGLang's aux/context capture never applies the target's final RMSNorm. verl-SpeCo receives the aux hidden and only reshapes it:

- `verl_speco/integration/sglang_patch.py` `_normalize_sglang_dflash_aux_hidden_states` (~L1829): reshape/permute/concat only, no norm.
- aux capture index is `target_layer_id + 1` (`verl_speco/integration/vllm_runtime.py` `[layer_id + 1 for layer_id in target_layer_ids]`), so `L-1` maps to the raw residual at the final capture point (before the final norm).

By contrast:

- **AutoModel** `HFDSparkTargetModel._feature` maps `layer_id == last` to `captured["norm"]` (post-norm), and `-1` to the embedding output (`nemo_automodel/components/speculative/dspark/target.py`, `.../dspark/common.py`).
- **verl-SpeCo old-logprob** `_resolve_hidden_state` maps `L-1` to `hidden_states[num_layers]` (HF appends the state after `self.norm` → post-norm), matching AutoModel for the last layer (`verl_speco/integration/oldlogprob_runtime.py:108-125`).

## Divergence table (per `target_layer_id` × collection path)

| layer_id | AutoModel (offline) | verl-SpeCo old-logprob | verl-SpeCo SGLang (aux) |
| --- | --- | --- | --- |
| intermediate `i` | raw layer-`i` output | raw layer-`i` output | raw layer-`i` output (consistent) |
| **last `L-1`** | **post-norm** | **post-norm** | **raw pre-norm** (divergent) |
| `-1` | **embedding** | **post-norm/last** (divergent) | not used in any yaml |

The separate post-norm "last_hidden" (the frozen `lm_head` input used for target logits) is **post-norm in all three paths**, so the correctness-critical logit path is fine.

## Impact

- **Default configs are unaffected.** Default `target_layer_ids` are `[2, L//2, L-3]` (EAGLE3) / `[1 … L-3]` / `[L//2]` (DFlash/DSpark) and never include `L-1` or `-1`, so all three paths agree.
- **It bites when a config puts the last layer in `target_layer_ids`** — which the AutoModel DSpark reference recipes do:
  - `examples/speculative/dspark/deepseek_v4_flash_dspark.yaml` → `[40, 41, 42]` on a 43-layer target (`42 = L-1`)
  - `examples/speculative/dspark/glm_5.2_dspark.yaml` → `[75, 76, 77]` on a 78-layer target (`77 = L-1`)
  For these, the last `fc` context slot is post-norm offline / via old-logprob but raw pre-norm via SGLang, so a drafter trained offline (or via old-logprob) and then collected/served via SGLang sees a mismatched feature; mixing `collect_hidden_states_from_sgl` and `collect_hidden_states_from_old_logprob` is likewise inconsistent. (Train-and-serve entirely on SGLang is self-consistent — pre-norm on both sides.)

## Proposed fix

Make the last-layer aux slot consistent across paths (post-norm), so drafters are portable between offline/old-logprob/SGLang. As a safe first step (SGLang aux lives deep in monkeypatched internals and cannot re-derive post-norm without model access), **fail closed** when SGLang hidden collection is combined with a resolved `target_layer_ids` that includes the last layer, directing users to old-logprob collection or to exclude `L-1`, with an explicit opt-out for the self-consistent SGLang-only train+serve case. A follow-up can apply the final norm to the last-layer aux slice to fully unify the paths. Also align the `-1` semantics (embedding vs post-norm-last) or reject `-1` explicitly.

AI assistance was used to investigate this.

### PR DESCRIPTION
**Pull request by khazic**
Closes #20.

## Problem

SGLang's aux/context capture never applies the target's final norm (`_normalize_sglang_dflash_aux_hidden_states` only reshapes; aux index is `target_layer_id + 1`). So a `target_layer_id` equal to the target's **last layer** (`num_hidden_layers - 1`) is captured as the raw **pre-norm** residual, while both the AutoModel offline reference and verl-SpeCo's own **old-logprob** collection capture it **post-norm** (HF appends the state after `self.norm`); `-1` likewise means the embedding offline but the post-norm last hidden in old-logprob.

Default `target_layer_ids` (`[2, L//2, L-3]` / `[1 … L-3]` / `[L//2]`) never hit `L-1` or `-1`, so defaults are unaffected. But the AutoModel DSpark reference recipes put the last layer in the context set (`[40,41,42]` on a 43-layer target, `[75,76,77]` on 78), so a drafter trained offline / via old-logprob and then collected or served via SGLang would see a mismatched `fc` feature for that slot; mixing `collect_hidden_states_from_sgl` with `collect_hidden_states_from_old_logprob` is likewise inconsistent. (Training and serving entirely on SGLang is self-consistent — pre-norm on both sides.) See #20 for the full three-way divergence table.

## Fix

Fail closed on the divergent combination (mirroring the existing SPECO validation style), without touching the fragile SGLang capture internals:

- `assert_sglang_aux_last_layer_norm_safe(layer_ids, num_hidden_layers, *, collect_from_sgl, allow_prenorm_last)` raises when SGLang collection is on and the resolved `target_layer_ids` include the last layer or `-1`.
- `SpecoRayPPOTrainer._speco_validate_sglang_aux_last_layer_norm` resolves the layer ids (via the same `resolve_oldlogprob_aux_layer_ids`) and the target depth, and runs the check once at collection setup. Best-effort: skips silently when the ids or depth cannot be resolved.
- New opt-out `actor_rollout_ref.rollout.drafter.training.allow_sglang_prenorm_last_layer` (default `false`) for the self-consistent SGLang-only train+serve case.

The error message directs users to (a) `collect_hidden_states_from_old_logprob=true` (which handles the last layer post-norm), (b) drop the last layer / `-1` from `target_layer_ids`, or (c) opt in.

## Testing

`tests/integration/test_sglang_last_layer_norm_guard.py`, 8 passed: intermediate layers pass; the last layer (the deepseek_v4 / glm_5.2 pattern) and `-1` are rejected; the opt-out and the non-SGLang path bypass; unresolvable inputs skip. Full suite `96 passed, 2 skipped, 2 failed` (both failures pre-exist on `main`, unrelated).

```
pytest tests/integration/test_sglang_last_layer_norm_guard.py -q
# 8 passed
```

## Known follow-up

A full fix would apply the target's final norm to the last-layer aux slice on the SGLang path (so all paths are post-norm-consistent and drafters are portable), rather than fail closed. That touches the monkeypatched SGLang capture and needs a live-SGLang GPU run to validate, so it is left as a follow-up; this PR closes the silent-mismatch path first.

### CODE REVIEW

**Comment by khazic** [comment]
Thanks, both valid and fixed in 819de5f:

**`-1` without a resolvable depth.** The guard returned early when `num_hidden_layers` was None, so a `target_layer_ids` containing `-1` (the embedding, divergent regardless of depth) slipped through. Now `-1` is rejected unconditionally and only the last-layer check is gated on a known depth. Added a regression test for the `-1` + None-depth case.

**Validate at init, not mid-rollout.** Moved `_speco_validate_sglang_aux_last_layer_norm` from `_speco_collect_generation_samples` (which runs only after the first, expensive rollout) to `init_workers`, right after `_init_speco_drafter_workers()`, so an invalid config fails closed before any generation. Dropped the now-unneeded one-shot flag.
